File: modules/context-native/ios/InboxAcknowledgementStore.swift
import Darwin
import Foundation

enum InboxAcknowledgementStoreError: Error, Equatable {
  case invalidIdentifier
  case integrityFailed
  case writeFailed
  case recoveryRequired

  var stableCode: String {
    switch self {
    case .invalidIdentifier: "SCHEMA_INVALID"
    case .integrityFailed: "ARTIFACT_INTEGRITY_FAILED"
    case .writeFailed: "STORAGE_WRITE_FAILED"
    case .recoveryRequired: "PIPELINE_RECOVERY_REQUIRED"
    }
  }
}

/**
 Durable, metadata-only evidence that an ingestion ID was already handed off and ACKed.

 The receipt is published before the scanner-visible Inbox directory is removed. It preserves
 the already-validated manifest, but never provider URLs, display filenames, or artifact bytes.
 */
enum InboxAcknowledgementStore {
  private static let directoryName = "InboxAcknowledgements"
  private static let tombstoneDirectoryName = "InboxAckTombstones"
  private static let maximumReceiptBytes = 262_144

  static func read(
    container: URL,
    ingestionId: String
  ) throws -> [String: Any]? {
    guard canonicalUUID(ingestionId) else {
      throw InboxAcknowledgementStoreError.invalidIdentifier
    }
    let root = container.appendingPathComponent(directoryName, isDirectory: true)
    guard FileManager.default.fileExists(atPath: root.path) else { return nil }
    try requireSafeDirectory(root, expectedParent: container)
    let receipt = root.appendingPathComponent("\(ingestionId).json")
    guard FileManager.default.fileExists(atPath: receipt.path) else { return nil }
    try requireReceiptPath(receipt, root: root)
    do {
      return try InboxManifestValidator.readAcknowledgementReceipt(
        receipt,
        ingestionId: ingestionId
      )
    } catch let error as InboxManifestValidationError {
      if error == .unsupportedVersion { throw error }
      throw InboxAcknowledgementStoreError.integrityFailed
    } catch {
      throw InboxAcknowledgementStoreError.integrityFailed
    }
  }

  static func publish(
    container: URL,
    ingestionId: String,
    manifestData: Data,
    directorySynchronizer: (URL) throws -> Void
  ) throws -> [String: Any] {
    guard canonicalUUID(ingestionId) else {
      throw InboxAcknowledgementStoreError.invalidIdentifier
    }
    guard !manifestData.isEmpty, manifestData.count <= maximumReceiptBytes else {
      throw InboxAcknowledgementStoreError.integrityFailed
    }
    let root = container.appendingPathComponent(directoryName, isDirectory: true)
    try ensureDurableDirectory(
      root,
      expectedParent: container,
      directorySynchronizer: directorySynchronizer
    )
    let receipt = root.appendingPathComponent("\(ingestionId).json")
    if FileManager.default.fileExists(atPath: receipt.path) {
      try requireReceiptPath(receipt, root: root)
      let existing: Data
      do { existing = try Data(contentsOf: receipt, options: [.mappedIfSafe]) }
      catch { throw InboxAcknowledgementStoreError.integrityFailed }
      guard existing == manifestData else {
        throw InboxAcknowledgementStoreError.integrityFailed
      }
      guard let manifest = try read(container: container, ingestionId: ingestionId) else {
        throw InboxAcknowledgementStoreError.integrityFailed
      }
      return manifest
    }

    let partial = root.appendingPathComponent(
      ".\(ingestionId)-\(UUID().uuidString.lowercased()).partial"
    )
    defer { try? FileManager.default.removeItem(at: partial) }
    do {
      guard FileManager.default.createFile(atPath: partial.path, contents: nil) else {
        throw InboxAcknowledgementStoreError.writeFailed
      }
      let output = try FileHandle(forWritingTo: partial)
      do {
        try output.write(contentsOf: manifestData)
        try output.synchronize()
        try output.close()
      } catch {
        try? output.close()
        throw error
      }
      guard Darwin.rename(partial.path, receipt.path) == 0 else {
        throw InboxAcknowledgementStoreError.writeFailed
      }
      try directorySynchronizer(root)
      guard let manifest = try read(container: container, ingestionId: ingestionId) else {
        throw InboxAcknowledgementStoreError.integrityFailed
      }
      return manifest
    } catch let error as InboxAcknowledgementStoreError {
      throw error
    } catch {
      throw InboxAcknowledgementStoreError.writeFailed
    }
  }

  static func matchingTombstones(
    container: URL,
    ingestionId: String
  ) throws -> [URL] {
    guard canonicalUUID(ingestionId) else {
      throw InboxAcknowledgementStoreError.invalidIdentifier
    }
    let root = container.appendingPathComponent(tombstoneDirectoryName, isDirectory: true)
    guard FileManager.default.fileExists(atPath: root.path) else { return [] }
    try requireSafeDirectory(root, expectedParent: container)
    do {
      return try FileManager.default.contentsOfDirectory(
        at: root,
        includingPropertiesForKeys: [.isDirectoryKey, .isSymbolicLinkKey]
      ).filter { tombstoneIngestionId($0, root: root) == ingestionId }
        .sorted { $0.lastPathComponent < $1.lastPathComponent }
    } catch let error as InboxAcknowledgementStoreError {
      throw error
    } catch {
      throw InboxAcknowledgementStoreError.writeFailed
    }
  }

  static func tombstoneIngestionId(_ candidate: URL, root: URL) -> String? {
    guard candidate.pathExtension == "ack",
          candidate.deletingLastPathComponent().standardizedFileURL == root.standardizedFileURL,
          let values = try? candidate.resourceValues(
            forKeys: [.isDirectoryKey, .isSymbolicLinkKey]
          ),
          values.isDirectory == true,
          values.isSymbolicLink != true else { return nil }
    let stem = candidate.deletingPathExtension().lastPathComponent
    guard stem.count == 73 else { return nil }
    let separator = stem.index(stem.startIndex, offsetBy: 36)
    guard stem[separator] == "-" else { return nil }
    let suffixStart = stem.index(after: separator)
    let ingestionId = String(stem[..<separator])
    return canonicalUUID(ingestionId) && canonicalUUID(String(stem[suffixStart...]))
      ? ingestionId
      : nil
  }

  private static func ensureDurableDirectory(
    _ root: URL,
    expectedParent: URL,
    directorySynchronizer: (URL) throws -> Void
  ) throws {
    if FileManager.default.fileExists(atPath: root.path) {
      try requireSafeDirectory(root, expectedParent: expectedParent)
    } else {
      do {
        try FileManager.default.createDirectory(at: root, withIntermediateDirectories: false)
      } catch {
        throw InboxAcknowledgementStoreError.writeFailed
      }
      try requireSafeDirectory(root, expectedParent: expectedParent)
    }
    do {
      try directorySynchronizer(root)
      try directorySynchronizer(expectedParent)
    } catch {
      throw InboxAcknowledgementStoreError.writeFailed
    }
  }

  private static func requireSafeDirectory(
    _ directory: URL,
    expectedParent: URL
  ) throws {
    guard directory.deletingLastPathComponent().standardizedFileURL
            == expectedParent.standardizedFileURL,
          let values = try? directory.resourceValues(
            forKeys: [.isDirectoryKey, .isSymbolicLinkKey]
          ),
          values.isDirectory == true,
          values.isSymbolicLink != true else {
      throw InboxAcknowledgementStoreError.writeFailed
    }
  }

  private static func requireReceiptPath(_ receipt: URL, root: URL) throws {
    guard receipt.deletingLastPathComponent().standardizedFileURL == root.standardizedFileURL,
          let values = try? receipt.resourceValues(
            forKeys: [.isRegularFileKey, .isSymbolicLinkKey]
          ),
          values.isRegularFile == true,
          values.isSymbolicLink != true else {
      throw InboxAcknowledgementStoreError.integrityFailed
    }
  }

  private static func canonicalUUID(_ value: String) -> Bool {
    guard let identifier = UUID(uuidString: value) else { return false }
    return identifier.uuidString.lowercased() == value
  }
}

File: modules/context-native/android/src/main/java/com/aicontextpack/nativebridge/InboxAcknowledgementStore.kt
package com.aicontextpack.nativebridge

import android.system.Os
import android.system.OsConstants
import java.io.File
import java.io.FileOutputStream
import java.util.UUID

internal class InboxAcknowledgementStoreException(
  val stableCode: String,
) : Exception(stableCode)

/**
 * Durable, metadata-only evidence that an ingestion ID was already handed off and ACKed.
 *
 * The receipt is published before the scanner-visible Inbox directory is removed. It preserves
 * the already-validated manifest, but never provider URIs, display filenames, or artifact bytes.
 */
internal object InboxAcknowledgementStore {
  private const val directoryName = "InboxAcknowledgements"
  private const val tombstoneDirectoryName = "InboxAckTombstones"
  private const val maximumReceiptBytes = 262_144

  fun read(filesDir: File, ingestionId: String): Map<String, Any?>? {
    requireCanonicalUuid(ingestionId)
    val root = File(filesDir, directoryName)
    if (!root.exists()) return null
    requireSafeDirectory(root, filesDir)
    val receipt = File(root, "$ingestionId.json")
    if (!receipt.exists()) return null
    requireReceiptPath(receipt, root)
    return try {
      InboxManifestScanner.readAcknowledgementReceipt(receipt, ingestionId)
    } catch (error: NativeException) {
      throw InboxAcknowledgementStoreException(
        if (error.code == "SCHEMA_VERSION_UNSUPPORTED") error.code else "ARTIFACT_INTEGRITY_FAILED",
      )
    }
  }

  fun publish(
    filesDir: File,
    ingestionId: String,
    manifestBytes: ByteArray,
    directorySynchronizer: (File) -> Unit,
  ): Map<String, Any?> {
    requireCanonicalUuid(ingestionId)
    if (manifestBytes.isEmpty() || manifestBytes.size > maximumReceiptBytes) {
      throw InboxAcknowledgementStoreException("ARTIFACT_INTEGRITY_FAILED")
    }
    val root = File(filesDir, directoryName)
    ensureDurableDirectory(root, filesDir, directorySynchronizer)
    val receipt = File(root, "$ingestionId.json")
    if (receipt.exists()) {
      requireReceiptPath(receipt, root)
      val existing = try { receipt.readBytes() }
      catch (_: Exception) {
        throw InboxAcknowledgementStoreException("ARTIFACT_INTEGRITY_FAILED")
      }
      if (!existing.contentEquals(manifestBytes)) {
        throw InboxAcknowledgementStoreException("ARTIFACT_INTEGRITY_FAILED")
      }
      return read(filesDir, ingestionId)
        ?: throw InboxAcknowledgementStoreException("ARTIFACT_INTEGRITY_FAILED")
    }

    val partial = File(root, ".$ingestionId-${UUID.randomUUID()}.partial")
    try {
      FileOutputStream(partial).use { output ->
        output.write(manifestBytes)
        output.fd.sync()
      }
      Os.rename(partial.path, receipt.path)
      directorySynchronizer(root)
      return read(filesDir, ingestionId)
        ?: throw InboxAcknowledgementStoreException("ARTIFACT_INTEGRITY_FAILED")
    } catch (error: InboxAcknowledgementStoreException) {
      throw error
    } catch (_: Exception) {
      throw InboxAcknowledgementStoreException("STORAGE_WRITE_FAILED")
    } finally {
      runCatching { partial.delete() }
    }
  }

  fun matchingTombstones(filesDir: File, ingestionId: String): List<File> {
    requireCanonicalUuid(ingestionId)
    val root = File(filesDir, tombstoneDirectoryName)
    if (!root.exists()) return emptyList()
    requireSafeDirectory(root, filesDir)
    return try {
      (root.listFiles() ?: throw InboxAcknowledgementStoreException("STORAGE_WRITE_FAILED"))
        .filter { candidate ->
          tombstoneIngestionId(candidate, root) == ingestionId
        }
        .sortedBy { it.name }
    } catch (error: InboxAcknowledgementStoreException) {
      throw error
    } catch (_: Exception) {
      throw InboxAcknowledgementStoreException("STORAGE_WRITE_FAILED")
    }
  }

  fun tombstoneIngestionId(candidate: File, root: File): String? = runCatching {
    val name = candidate.name
    if (!name.endsWith(".ack")) return@runCatching null
    val stem = name.removeSuffix(".ack")
    if (stem.length != 73 || stem[36] != '-') return@runCatching null
    val ingestionId = stem.substring(0, 36)
    val nonce = stem.substring(37)
    if (!canonicalUuid(ingestionId) || !canonicalUuid(nonce)) return@runCatching null
    if (!candidate.isDirectory ||
      OsConstants.S_ISLNK(Os.lstat(candidate.path).st_mode) ||
      candidate.parentFile?.canonicalFile != root.canonicalFile
    ) return@runCatching null
    ingestionId
  }.getOrNull()

  private fun ensureDurableDirectory(
    root: File,
    filesDir: File,
    directorySynchronizer: (File) -> Unit,
  ) {
    try {
      if (root.exists()) {
        requireSafeDirectory(root, filesDir)
      } else {
        if (!root.mkdir()) throw InboxAcknowledgementStoreException("STORAGE_WRITE_FAILED")
        requireSafeDirectory(root, filesDir)
      }
      directorySynchronizer(root)
      directorySynchronizer(filesDir)
    } catch (error: InboxAcknowledgementStoreException) {
      throw error
    } catch (_: Exception) {
      throw InboxAcknowledgementStoreException("STORAGE_WRITE_FAILED")
    }
  }

  private fun requireSafeDirectory(directory: File, expectedParent: File) {
    try {
      if (!directory.isDirectory ||
        OsConstants.S_ISLNK(Os.lstat(directory.path).st_mode) ||
        directory.parentFile?.canonicalFile != expectedParent.canonicalFile
      ) {
        throw InboxAcknowledgementStoreException("STORAGE_WRITE_FAILED")
      }
    } catch (error: InboxAcknowledgementStoreException) {
      throw error
    } catch (_: Exception) {
      throw InboxAcknowledgementStoreException("STORAGE_WRITE_FAILED")
    }
  }

  private fun requireReceiptPath(receipt: File, root: File) {
    try {
      if (!receipt.isFile ||
        OsConstants.S_ISLNK(Os.lstat(receipt.path).st_mode) ||
        receipt.parentFile?.canonicalFile != root.canonicalFile
      ) {
        throw InboxAcknowledgementStoreException("ARTIFACT_INTEGRITY_FAILED")
      }
    } catch (error: InboxAcknowledgementStoreException) {
      throw error
    } catch (_: Exception) {
      throw InboxAcknowledgementStoreException("ARTIFACT_INTEGRITY_FAILED")
    }
  }

  private fun requireCanonicalUuid(value: String) {
    if (!canonicalUuid(value)) {
      throw InboxAcknowledgementStoreException("SCHEMA_INVALID")
    }
  }

  private fun canonicalUuid(value: String): Boolean = try {
    UUID.fromString(value).toString() == value
  } catch (_: IllegalArgumentException) {
    false
  }
}
